### DISCUSSION TITLE
Paper "Adaptive LL(*) Parsing" and code

### DISCUSSION THREAD

**Discussion by andr1972** [antlr/antlr4#4725] [General]
I have read "Adaptive LL(*) Parsing: The Power of Dynamic Analysis"
Are there any more articles on this topic? In which place of Antlr sources is core of this algorithm - equivalents of pseudocode functions from this article? Functions : parse, adaptivePredict,...

**Comment by kaby76** [comment]
* Source code for AdaptivePredict(): [CSharp](https://github.com/antlr/antlr4/blob/933c4e0a0d4d5accfc2f9fa416e2c0558b4512e0/runtime/CSharp/src/Atn/ParserATNSimulator.cs#L304-L386), [JavaScript](https://github.com/antlr/antlr4/blob/933c4e0a0d4d5accfc2f9fa416e2c0558b4512e0/runtime/JavaScript/src/antlr4/atn/ParserATNSimulator.js#L293-L361), [TypeScript in Antlr4ng](https://github.com/mike-lischke/antlr4ng/blob/6407a943d318ae4586e49101de14c3600e2ffb88/src/atn/ParserATNSimulator.ts#L316-L389), ...
* If you want to really understand the code, you have to do what we've been doing for many, many decades. Read and follow what it does in a debugger. There are also switches for tracing the parser (see [CSharp](https://github.com/antlr/grammars-v4/blob/199a5121ece05d2f2e7eca330d0738220499e80c/_scripts/templates/CSharp/st.Test.cs#L326), [Java](https://github.com/antlr/grammars-v4/blob/199a5121ece05d2f2e7eca330d0738220499e80c/_scripts/templates/Java/st.Test.java#L185)).
* The original. [Parr, T., Harwell, S., Fisher, K.: Adaptive LL(*) parsing: the power of dynamic analysis. In: Object-Oriented Programming, Systems, Languages, and Applications (2014).](https://dl.acm.org/doi/10.1145/2714064.2660202). [Free here](https://www.antlr.org/papers/allstar-techreport.pdf). Note, this paper is actually pretty good. The pseudocode is in the paper, and the implementation follows it closely.
* Oldie but goodie. [LL(*): the foundation of the ANTLR parser generator](https://dl.acm.org/doi/10.1145/1993316.1993548)
* A Coc-verified implementation of ALL(*), definitely worth a read. [Lasser, S., Casinghino, C., Fisher, K., Roux, C.: CoStar: a verified ALL(*) parser. In: Programming Language Design and Implementation (2021).](https://zenodo.org/records/4681598)
* CoStar++. [Lasser, S., Casinghino, C., Egolf, D., Fisher, K., Roux, C. (2023). Verified ALL(*) Parsing with Semantic Actions and Dynamic Input Validation. In: Rozier, K.Y., Chaudhuri, S. (eds) NASA Formal Methods. NFM 2023. Lecture Notes in Computer Science, vol 13903. Springer, Cham.](https://doi.org/10.1007/978-3-031-33170-1_25)
